File: src/com/snowballtech/algorithm/zhiyuancai/Week36.go
package main

import "fmt"

//定义单链表结构体
type ListNode struct {
	Val  int
	Next *ListNode
}

//思路:使用二路归并排序
func main() {
	// 改为控制台输入方式  Add by Max Yu 2021.09.09
	//要排序的数组
	arr := []int{-1, 5, 3, 4, 0}

	//头节点
	var head *ListNode = &ListNode{}
	var p *ListNode = head

	//数组放入链表
	for val := range arr {
		p.Next = &ListNode{Val: arr[val]}
		p = p.Next
	}

	//排序
	result := sort(head.Next)

	//输出
	fmt.Print("[")
	for true {
		if result != nil {
			fmt.Print(result.Val)
			result = result.Next
			if result != nil {
				fmt.Print(",")
			}
		} else {
			fmt.Print("]")
			break
		}
	}

}

//分割链表
func sort(node *ListNode) *ListNode {
	//若有该节点存在且有下一个节点
	if node == nil || node.Next == nil {
		return node
	}

	//快慢指针
	slow := node
	fast := node
	//查找中间节点
	for true {
		if fast.Next == nil || fast.Next.Next == nil {
			break
		} else {
			slow = slow.Next
			fast = fast.Next.Next
		}
	}
	//中间节点
	mid := slow.Next
	//左链表（中点之后赋空）
	slow.Next = nil

	//递归分割链表
	return merge(sort(node), sort(mid))
}

//二路归并
func merge(left *ListNode, right *ListNode) *ListNode {
	//结果
	result := &ListNode{}
	//结果游标
	temp := result

	//归并排序，将两个连表彰值较小的节点放入结果链表（不能直接赋予指针，会导致地址变换，返回结果为空）
	for true {
		if left == nil || right == nil {
			break
		}
		if left.Val < right.Val {
			temp.Val = left.Val
			left = left.Next
		} else {
			temp.Val = right.Val
			right = right.Next
		}
		//创建Next指针并移动游标
		temp.Next = &ListNode{}
		temp = temp.Next
	}

	//将剩余的链表直接赋值给结果链表（不能直接赋予指针，会导致地址变换，返回结果为空）
	if left != nil {
		temp.Val = left.Val
		temp.Next = left.Next
	}
	if right != nil {
		temp.Val = right.Val
		temp.Next = right.Next
	}
	return result
}

File: src/com/snowballtech/algorithm/zhiyuancai/Week36.java
package com.snowballtech.algorithm.zhiyuancai;

import java.util.*;

//思路:使用二路归并排序
public class Week36 {

    public static void main(String[] args) {
        //头结点
        ListNode head = new ListNode();

        //输入要排序的数
        ListNode next = null;
        System.out.println("请输入要排序的数 回车分割，输入end代表结尾");
        while(true){
            Scanner input = new Scanner(System.in);
            String s = input.next();
            if(s.equals("end")){
                break;
            }else {
                if(next == null){
                    next = new ListNode(Integer.parseInt(s));
                    head.next = next;
                } else {
                    next.next = new ListNode(Integer.parseInt(s));
                    next = next.next;
                }
            }
        }

        //排序
        ListNode sort = sort(head.next);

        //输出
        System.out.print("[");
        while (true) {
            System.out.print(sort.val);
            if (sort.next == null) {
                System.out.print("]");
                break;
            }
            sort = sort.next;
            if(sort != null){
                System.out.print(",");
            }
        }
    }

    //将链表分割为两份
    public static ListNode sort(ListNode head){
        //头结点为空或下一个节点为空时，结束递归
        if (head == null || head.next == null){
            return head;
        }

        //查找中点
        ListNode slow = head;
        ListNode quick = head;
        //快慢指针，快指针到末尾时，满指针到中点
        while(quick.next != null && quick.next.next != null){
            slow = slow.next;
            quick = quick.next.next;
        }
        //中间节点
        ListNode mid = slow.next;
        //截取左链表至中间节点
        slow.next = null;

        //递归左链表和右链表，然后进行合并
        return merge(sort(head), sort(mid));
    }

    //合并左右链表
    public static ListNode merge(ListNode left, ListNode right){
        //排序结果的头结点
        ListNode result = new ListNode();
        //排序结果
        ListNode p = result;

        //循环左链表和右链表，直到一方链表结束
        while (left != null && right != null){
            //将值较小的节点放入结果链表中
            if (left.val < right.val){
                p.next = left;
                left = left.next;
            }else {
                p.next = right;
                right = right.next;
            }
            //移动链表指针
            p = p.next;
        }

        //若右链表结束，则将左链表放入结果链表
        if (left != null){
            p.next = left;
        }
        //若左链表结束，则将右链表放入结果链表
        if (right != null){
            p.next = right;
        }

        return result.next;
    }
}
